### PR TITLE
Login feature

### PR DESCRIPTION
**Pull request by PedroAGC**
[INT-2977](https://jira.bigcommerce.com/browse/INT-2977)

## What?
Add **Authorize.Net** recommended banner for **Evo Referral Id** store.

## Why?
This helps merchants to setup Authorize.Net more easily. I changed another provider module because a store should display only two banners instead of three banners

## Testing / Proof
[https://drive.google.com/file/d/1QYYD1_c82Lp32scf162A_HgmZHWvLgUL/view?usp=sharing](https://drive.google.com/file/d/1QYYD1_c82Lp32scf162A_HgmZHWvLgUL/view?usp=sharing&authuser=0)

## How can this change be undone in case of failure?
1. Revert this PR

ping @bigcommerce/apex-integrations @bigcommerce/checkout

### CODE REVIEW

File: src/controllers/auth.controller.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import bcrypt from 'bcryptjs';
+import User from '../models/user.model';
+import generateJWT from '../helpers/jwtGenerator';
+
+const login = async (email: string, password: string) => {
+    //  Verify if the user exists in DB
+    const user = await User.findOne({where: {email: email}})
+    if(!user)
+        throw new Error('Invalid Credentials');
```
Comment (christianahvilla) [inline src/controllers/auth.controller.ts]:
```suggestion
    if(!user){
        throw new Error('Invalid Credentials');
    }
```

File: src/controllers/auth.controller.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    //  Verify the password
+    const validPassword =  bcrypt.compareSync(password, user.password)
+    if(!validPassword)
+        throw new Error('Invalid Credentials');
```
Comment (christianahvilla) [inline src/controllers/auth.controller.ts]:
```suggestion
    if(!validPassword) {
        throw new Error('Invalid Credentials');
     }
```

File: src/controllers/auth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    const token = await generateJWT(user.id);
+
+    const data = {
+        'user': user.email,
+        'token': token,
+    };
+    return data;
```
Comment (christianahvilla) [inline src/controllers/auth.controller.ts]:
```suggestion

    return data;
```

File: src/controllers/user.controller.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import bcrypt from 'bcryptjs';
+import { v4 as uuidv4  } from 'uuid';
+
+import User, { UserInput, UserOutput } from '../models/user.model';
```
Comment (christianahvilla) [inline src/controllers/user.controller.ts]:
```suggestion
import User, { UserInput, UserOutput } from '../models/user.model';
```

File: src/controllers/user.controller.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    } = payload;
+    //  Verify if the user email exists
+    const userExists = await User.findOne({where: {email: email}})
+    if(userExists){
+        throw new Error(`The user with the email ${email} is already registered`);
+    }
+    const phoneExists = await User.findOne({where: {phone: phone}})
+    if(phoneExists){
+        throw new Error(`The user with the phone ${phone} is already registered`);
+    }
```
Comment (christianahvilla) [inline src/controllers/user.controller.ts]:
```suggestion
    if(userExists) {
        throw new Error(`The user with the email ${email} is already registered`);
    }
    const phoneExists = await User.findOne({where: {phone: phone}})
    if(phoneExists) {
        throw new Error(`The user with the phone ${phone} is already registered`);
    }
```

File: src/database/connection.ts [hunk 6]
```diff
@@ -1 +1,14 @@
...
+    config.DB_DATABASE,
+    config.DB_USERNAME,
+    config.DB_PASSWORD, {
+        host: config.DB_HOST,
+        dialect: 'mysql',
+    },
+
+);
+
+export default db;
```
Comment (christianahvilla) [inline src/database/connection.ts:15]: add new line after end of file

File: src/helpers/jwtGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+        jwt.sign(payload, config.JWT_KEY, {
+            //expiresIn: '4hr'
+        }, (err, token) => {
+            if(err){
+                console.log(err);
+                reject('The ');
+            }else{
+                resolve(token);
+            }
```
Comment (christianahvilla) [inline src/helpers/jwtGenerator.ts]:
```suggestion
        }, (err, token) => {
            if(err) {
                console.log(err);
                reject('The ');
            } else {
                resolve(token);
            }
```

File: src/middlewares/fieldsValidator.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { Request, Response, NextFunction } from 'express';
+import { validationResult } from 'express-validator';
+
+const fieldsValidation = (req: Request, res: Response, next: NextFunction ) => {
+    const errors =  validationResult(req);
+    if(!errors.isEmpty())
+        return res.status(400).json(errors);
```
Comment (christianahvilla) [inline src/middlewares/fieldsValidator.ts]:
```suggestion
    if (!errors.isEmpty()) {
    
        return res.status(400).json(errors);
    }
```

File: src/middlewares/fieldsValidator.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    next();
+}
+
+export default fieldsValidation;
```
Comment (christianahvilla) [inline src/middlewares/fieldsValidator.ts:13]:
```suggestion
export default fieldsValidation;

```

File: src/models/index.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import User from "./user.model";
+
+
```
Comment (christianahvilla) [inline src/models/index.ts]:
```suggestion

```

File: src/models/user.model.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    public last_name!: string;
+    public email!: string;
+    public phone!: string;
+    public role!: string;
+    public email_verified_at!: Date;
+    public password!: string;
+    public readonly createdAt!: Date;
+    public readonly updatedAt!: Date;
+    public readonly deletedAt!: Date;
+    dataValues: any;
```
Comment (christianahvilla) [inline src/models/user.model.ts]: why is it neccesary

File: src/models/user.model.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+User.init({
```
Comment (christianahvilla) [inline src/models/user.model.ts]: what is it?

File: src/models/user.model.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        allowNull: true,
+    }
+},
+{
+    timestamps: true,
+    sequelize: db,
+    paranoid: true
+});
+
+export default User;
```
Comment (christianahvilla) [inline src/models/user.model.ts:103]:
```suggestion
export default User;

```

File: src/routes/auth.routes.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import { Router, Request, Response } from 'express';
+import { check } from 'express-validator';
+
+
+import {login} from '../controllers/auth.controller';
+import fieldsValidation from '../middlewares/fieldsValidator';
```
Comment (christianahvilla) [inline src/routes/auth.routes.ts]:
```suggestion
import {login} from '../controllers/auth.controller';
import fieldsValidation from '../middlewares/fieldsValidator';
```

File: src/routes/auth.routes.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    try {
+        const data = await login(email, password);
+
+        res.status(200).json(data)
+    } catch (error: any) {
+        res.status(400).json({msg: error.message});
+    }
+})
+
+export default router;
```
Comment (christianahvilla) [inline src/routes/auth.routes.ts:24]:
```suggestion
export default router;

```

File: src/routes/user.routes.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
+import {Router, Request, Response} from 'express'
+import { check } from 'express-validator';
+
+import fieldsValidation from '../middlewares/fieldsValidator';
```
Comment (christianahvilla) [inline src/routes/user.routes.ts:4]:
```suggestion
import fieldsValidation from '../middlewares/fieldsValidator';
```

File: src/routes/user.routes.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    check('phone', 'Is not a valid phone number').isLength({min: 7}),
+    check('password', 'Is not a valid password').isLength({min: 6}),
+    fieldsValidation
+], async(req: Request, res: Response) => {
+
+    try {
+        const user = await create(req.body);
+        //delete
+
+        res.status(201).json(user);
```
Comment (christianahvilla) [inline src/routes/user.routes.ts]: is it necessary?

File: src/server.ts [hunk 18]
```diff
@@ -1,23 +1,44 @@
 import express, { Application } from 'express';
+
 import cors from 'cors';
 
+import config from './config/config';
+import authRoutes from './routes/auth.routes';
+import userRoutes from './routes/user.routes';
+import dbInit from './database/init';
```
Comment (christianahvilla) [inline src/server.ts]:
```suggestion
import express, { Application } from 'express';
import cors from 'cors';
import config from './config/config';
import authRoutes from './routes/auth.routes';
import userRoutes from './routes/user.routes';
import dbInit from './database/init';
```

File: src/server.ts [hunk 19]
```diff
@@ -28,4 +49,4 @@ class Server{
 
 }
 
-export default Server;
\ No newline at end of file
+export default Server;
```
Comment (christianahvilla) [inline src/server.ts:49]:
```suggestion
export default Server;

```

**Review by christianahvilla — CHANGES_REQUESTED** [review]
Please check unnecessary comments, add comment when it is necesary, add new line at the of files.

**Comment by PedroAGC** [comment]
@christianahvilla  Fixed!

**Review by christianahvilla — APPROVED** [review]
LGTM

**Review by christianahvilla — APPROVED** [review]
LGTM

**Comment by christianahvilla** [comment]
Please rebase your branch @PedroAGC

**Pull request merged**